help/gjiten.1
.TH GJITEN 1 "20 Jun 2002"
.SH NAME
\fBgjiten\fP \- Japanese dictionary for Gnome
.SH SYNOPSIS
.B gjiten [ option ]

.SH DESCRIPTION
.B gjiten
is a Japanese dictionary for the Gnome Desktop. It also has a kanji dictionary; 
any combination of stroke number, radicals and search key can be used for kanji lookups. 
Requires dictionary files in edict format and a working X Input Method (e.g. kinput2) for Japanese input.

This manpage only explains the command line options. You should be able to access the complete documentation from the Help menu.
.LP
.SH OPTIONS
\fB\-c\fR and \fB\-v\fR are convenient for fast lookups via keyboard shortcuts.

.TP
\fB\-k\fR
Same as \fB\-\-kanjidic\fR
.TP
\fB\-\-kanjidic\fR
Only show the \fBKanjiDic\fR window on startup.
.TP
\fB\-w WORD\fR
Same as \fB\-\-word\-lookup=WORD\fR
.TP
\fB\-\-word\-lookup=WORD\fR
Look up \fBWORD\fR on startup.
.TP
\fB\-l KANJI\fR
Same as \fB\-\-kanji\-lookup=KANJI\fR
.TP
\fB\-\-kanji\-lookup=KANJI\fR
Look up \fBKanji\fR on startup in \fBKanjiDic\fR.
.TP
\fB\-c\fR
Same as \fB\-\-clip\-kanji\fR
.TP
\fB\-\-clip\-kanji\fR
Look up kanji from the \fBclipboard\fR.
.TP
\fB\-v\fR
Same as \fB\-\-clip\-word\fR
.TP
\fB\-\-clip\-word\fR
Look up word from the \fBclipboard\fR.
.TP
\fB\-\-version\fR
Prints the version of gjiten

.PB
.SH BUGS
See the BUGS file for a list of bugs. The TODO file is more likely to contain these. Please contact me if you encounter any bugs.

.SH SEE ALSO
.RB xjdic(1)

.SH AUTHORS
Botond Botyanszki (boti@rocketmail.com)


ChangeLog
v2.5 ()
    - History buffer overrun fixed (John Thacker).
    

v2.4 (06/05/2005)
    - Gettext fix for kanji info pref dialog. (Daichi Kawahata)
    - KanjiDic segv on CTRL-Q fixed. (Alexis S. L. Carvalho)
    - Japanese translation fixes and desktop file i18n. (Daichi Kawahata)
    - Dicfile match colorization fix for gettext translation.
    - KanjiDic history width problem fixed.
    - Added support for the Chinese CEDICT dictionary.
    - German translation update. (Gregor Riepl)
    - Optional unicode radicals in KanjiDic.
    - Don't limit if result limit is < 1.
    - Perl script for automated dictionary setup (system-wide or user-only).
    - Gconf schema for system-wide settings.
    - Fixed help menu.

v2.3 (10/06/2004)
    - Fixed segv when kanjipad entry is NULL.
    - Made search result textviews non-editable.
    - Segv after dicfile preferences change fixed.
    - Added Show/Hide search options button to the toolbar.

v2.2 (06/13/2004)
    - Search hiragana on katakana input and vice versa.
    - Major code cleanup.
    - Fix 'whole expression' matching with newer edict files (ignore metadata in parenthesis).
    - Search entry receives focus (and input) on keypress if in focus.
    - Accelerator keys.
    - Preferences rewritten, is now glade based.
    - Unlimited number of dictionary files.
    - radkfile updated.

v2.1 (05/05/2003)
    - GConf crash fixed.
    - Set icon with GNOME.
    - German translation.

v2.0 (01/04/2003)
    - Almost full rewrite:
    - xjdic code removed, no more index files.
    - Now uses GConf for storing configuration
    - Ported to the GNOME 2.0 platform.
    - Enhanced verb deinflection.
    - Full unicode support, dictionary files need to be utf-8.
    - Manual updated, now works with GNOME2 help system.
    
v1.0 (01/19/2002)
    - "Reply overrun" crashes with the JP-German GDICT fixed.
    - Search options are saved on exit.
    - Limit search results in word dictionary. 
    - Kanji info window moved to the main KanjiDic window. 
    - KanjiDic kanji-info history.
    - Automatic vertical scrollbars.
    - Bigger fonts for KanjiDic. 
    - Word dictionary fonts also customizable in preferences. 
    
v0.9 (10/25/2001)
    - History saving (up to 50 entries) on exit. 
    - Auto adjust search options if no match is found.
    - Now starts up properly with gnome/gtk specific cmd line options.
    - Updated for the IA64 architecture. 
    - Shouldn't crash with Fvwm2 anymore.
    - dpkg-buildpackage should work under debian potato from now.
    - Dicfile limit raised to 40.
    - Added Back/Forward buttons to the toolbar.
    - Radkfile updated.    
    - Fixed history behaviour (garbage, whitespace, case sensitivity)
    - Segfault on unconfigured exit fixed.
    - Numerous small bugfixes.
        
v0.8 (05/01/2001)
    - Debianized source. =>Debian packages.
    - Documentation updated and converted to DocBook sgml.
    - Makefile updates.
    - Japanese localization fixes.
    - Search corrections.            
    - Real time kanji lookup.
    - Clipboard lookup (kanji&word) via command line. You can assign 
       a keyboard shortcut with your window manager for fast lookups. 
    - Docs moved to /usr/share/doc 
    - Click on radical removes it from the list if already there.
    - Fixed memory leaks.
    - Popup window behaviour fixed.    
    
v0.7 (11/03/2000)
    - Various small fixes, code cleanups.
    - It's more user friendly on first startup now.
    - $prefix is correctly substituted in source.
    - Kanji info window behaviour fixed.        
    - Radical selection window rewritten to use buttons.
    - 'No dicfiles crash' fixed.
    - Command line arguments (word and kanji lookup). See gjiten --help.    
    - Gnome menu entry under Applications (gjiten.desktop file).
    - Linear Search fixed.
                
v0.6 (05/21/2000)
    - Linear search added: slower but more accurate. 
    - Makefile fixes for compiling under Debian.
    - Documentation updates (both ja and en). 
    - Should be more tolerant with file errors.    
    - Japanese localization: ja.po added. Set LANGUAGE=ja for Japanese messages.
    - Copy, paste buttons.
    - Lots of other fixes.

v0.5 (04/19/2000)	
    - Togglebuttons in Pref/kanjidic should work okay now. 
    - Widget shading in KanjiDic via checkbuttons.
    - README converted to html, added as Help/help.
    - Set LC_ALL to ja_JP too. 
    - Added "Search All Dics" feature.
    - Logo.
    - Documentation updates.
    - Automake/Autoconf/Install fixes.
    - Appbar messages.
    - Code reorganized, misc fixes.
    - Added Japanese documentation.
        
v0.4 (03/02/2000)
    - Verb deinflection works now.
    - Small cleanups and fixes. Preferences changed.
    - Config file changed to GNOME style.
    - Dictionary selection button changed to pulldown menu.
    - KanjiPad startup button added. 
    - Text Entry widgets replaced with Combo Entry widgets [search history]. 
    -"Extra character after highlighted text" bug fixed.

v0.3 (01/09/2000)
    -Added an option for using more than one dictionary file.
    -Added missing radkfile.
    
v0.2 (12/28/1999)
    -Added a Kanji dictionary with some nice features.
    -Added more options to preferences.
    -Word dictionary lookup functions rewritten. Verb inflection feature is not 
     implemented yet.
    -More bugs, known ones killed ;)
       
v0.1 (12/25/1999) 
    -First release. 


src/gjiten.c
/* -*- Mode: C; tab-width: 2; indent-tabs-mode: t; c-basic-offset: 2 -*- */
/* gjiten.c

   GJITEN : A GTK+/GNOME BASED JAPANESE DICTIONARY

   Copyright (C) 1999 - 2003 Botond Botyanszki <boti@rocketmail.com>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published  by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <string.h>
#include <libgnome/libgnome.h>
#include <libgnomeui/libgnomeui.h>
#include <gconf/gconf-client.h>
#ifdef HAVE_INTTYPES_H
#include <inttypes.h>
#endif
#ifdef HAVE_STDINT_H
#include <stdint.h>
#endif

#include "error.h"
#include "constants.h"
#include "conf.h"
#include "dicfile.h"
#include "worddic.h"
#include "kanjidic.h"
#include "gjiten.h"
#include "dicutil.h"

GjitenApp *gjitenApp;

static void parse_an_arg(poptContext state,
												 enum poptCallbackReason reason,
												 const struct poptOption *opt,
												 const char *arg, void *data);

/***************** VARIABLES ***********************/

gchar *clipboard_text = NULL;

/* FIXME: GConfEnumPair also in kanjidic.h */
gchar *kanjidicstrg[] = { "kanji", "radical", "strokes", "reading", "korean", 
													"pinyin", "english", "bushu", "classic", "freq", "jouyou",
													"deroo", "skip", "fourc", "hindex", "nindex", "vindex",
													"iindex", "mnindex", "mpindex", "eindex", "kindex", 
													"lindex", "oindex", "cref", "missc", "unicode", "jisascii" };


enum {
  KANJIDIC_KEY        = -1,
  WORD_LOOKUP_KEY     = -2,
  KANJI_LOOKUP_KEY    = -3,
  CLIP_KANJI_KEY      = -4,
  CLIP_WORD_KEY       = -5
};

/* Command line arguments via popt */
static struct poptOption arg_options [] = {
  { NULL, '\0', POPT_ARG_CALLBACK, (gpointer)parse_an_arg, 0,
    NULL, NULL },

  { "kanjidic", 'k', POPT_ARG_NONE, NULL, KANJIDIC_KEY,
    N_("Start up Kanjidic instead of Word dictionary"), NULL },

  { "word-lookup", 'w', POPT_ARG_STRING, NULL, WORD_LOOKUP_KEY,
    N_("Look up WORD in first dictionary"), N_("WORD") },

  { "kanji-lookup", 'l', POPT_ARG_STRING, NULL, KANJI_LOOKUP_KEY,
    N_("Look up KANJI in kanji dictionary"), N_("KANJI") },

  { "clip-kanji", 'c', POPT_ARG_NONE, NULL, CLIP_KANJI_KEY,
    N_("Look up kanji from clipboard"), NULL },

  { "clip-word", 'v', POPT_ARG_NONE, NULL, CLIP_WORD_KEY,
    N_("Look up word from clipboard"), NULL },

  { NULL, '\0', 0, NULL, 0, NULL, NULL }

};



/*================ Functions ===============================================*/

static void parse_an_arg(poptContext state,
												 enum poptCallbackReason reason,
												 const struct poptOption *opt,
												 const char *arg, void *data) {

  
  switch (opt->val) {
  case KANJIDIC_KEY:
    gjitenApp->conf->startkanjidic = TRUE;
    break;
  case WORD_LOOKUP_KEY:
    gjitenApp->conf->word_to_lookup = (gchar *)arg;
    break;
  case KANJI_LOOKUP_KEY:
    gjitenApp->conf->kanji_to_lookup = (gchar *)arg;
    break;
  case CLIP_KANJI_KEY:
    gjitenApp->conf->clip_kanji_lookup = TRUE;
    gjitenApp->conf->clip_word_lookup = FALSE;
    break;
  case CLIP_WORD_KEY:
    gjitenApp->conf->clip_word_lookup = TRUE;
    gjitenApp->conf->clip_kanji_lookup = FALSE;
    break;
  default:
    break;
  }
}

void gjiten_clear_entry_box(gpointer entrybox) {
  gtk_entry_set_text(GTK_ENTRY(entrybox), "");
}

void gjiten_exit() {
	if ((gjitenApp->worddic == NULL) && (gjitenApp->kanjidic == NULL)) {
		GJITEN_DEBUG("gjiten_exit()\n");
		conf_save_options(gjitenApp->conf);
		dicutil_unload_dic();
		dicfile_list_free(gjitenApp->conf->dicfile_list);
		conf_close_handler();
		gtk_main_quit();
	}
}

void gjiten_start_kanjipad() {
  FILE *kanjipad_binary;
  char *kpad_cmd;

  kanjipad_binary = fopen(gjitenApp->conf->kanjipad,"r");
  if (kanjipad_binary == NULL) {
    gjiten_print_error(_("Couldn't find the KanjiPad executable!\n"
												 "Please make sure you have it installed on your system \n"
												 "and set the correct path to it in the Preferences.\n"
												 "See the Documentation for more details about KanjiPad."));
  }
  else {
    fclose(kanjipad_binary);
    kpad_cmd = g_malloc(strlen(gjitenApp->conf->kanjipad) + 2);
    strcpy(kpad_cmd, gjitenApp->conf->kanjipad);
    strcat(kpad_cmd, "&");
    system(kpad_cmd); /* FIXME */
    g_free(kpad_cmd);
  }
}

void gjiten_display_manual(GtkWidget *widget, void *data) {
  GtkWidget *window = data;
  GError *err = NULL;
  gboolean retval = FALSE;

  retval = gnome_help_display("gjiten.xml", NULL, &err);
  
  if (retval == FALSE) {
    GtkWidget *dialog;
    dialog = gtk_message_dialog_new(GTK_WINDOW(window),
																		GTK_DIALOG_DESTROY_WITH_PARENT,       
																		GTK_MESSAGE_ERROR,
																		GTK_BUTTONS_CLOSE,
																		_("Could not display help: %s"),
																		err->message);
    
    g_signal_connect(G_OBJECT(dialog), "response",
										 G_CALLBACK(gtk_widget_destroy),
										 NULL);
    
    gtk_window_set_resizable(GTK_WINDOW(dialog), FALSE);
    
    gtk_widget_show(dialog);
    
    g_error_free(err);
  }
}


void gjiten_create_about() {
  const gchar *authors[] = { "Botond Botyanszki <boti@rocketmail.com>", NULL };
  const gchar *documenters[] = { NULL };
  const gchar *translator = _("TRANSLATORS! PUT YOUR NAME HERE");
  static GtkWidget *about = NULL;
  GdkPixbuf *pixbuf = NULL;

  if (about != NULL) {
    gdk_window_show(about->window);
    gdk_window_raise(about->window);
    return;
  }

  if (pixbuf != NULL) {
    GdkPixbuf* temp_pixbuf = NULL;
    
    temp_pixbuf = gdk_pixbuf_scale_simple(pixbuf, 
					   gdk_pixbuf_get_width(pixbuf) / 2,
					   gdk_pixbuf_get_height(pixbuf) / 2,
					   GDK_INTERP_HYPER);
    g_object_unref(pixbuf);
    pixbuf = temp_pixbuf;
  }

  pixbuf = gdk_pixbuf_new_from_file(PIXMAPDIR"/gjiten/gjiten-logo.png", NULL);

  if (strcmp(translator, "translated_by") == 0) translator = NULL;

  /*
    _("Released under the terms of the GNU GPL.\n"
    "Check out http://gjiten.sourceforge.net for updates"), 
  */
  about = gnome_about_new("gjiten", VERSION, "Copyright \xc2\xa9 1999-2005 Botond Botyanszki",
			  _("gjiten is a Japanese dictionary for Gnome"),
			  (const char **)authors,
			  (const char **)documenters,
			  (const char *)translator,
			  pixbuf);

  gtk_window_set_destroy_with_parent(GTK_WINDOW(about), TRUE);
  if (pixbuf != NULL)  g_object_unref (pixbuf);

  g_signal_connect(G_OBJECT(about), "destroy", G_CALLBACK(gtk_widget_destroyed), &about);
  gtk_widget_show(about);

}



/*********************** MAIN ***********************************/
int main (int argc, char **argv) {
  char *icon_path = PIXMAPDIR"/jiten.png";
 
	gjitenApp = g_new0(GjitenApp, 1);
  conf_init_handler();
	gjitenApp->conf = conf_load();

	if (gjitenApp->conf->envvar_override == TRUE) {
		if (gjitenApp->conf->gdk_use_xft == TRUE) putenv("GDK_USE_XFT=1");
		else putenv("GDK_USE_XFT=0");
		/* if (gjitenApp->conf->force_ja_JP == TRUE) putenv("LC_CTYPE=ja_JP"); */
		if (gjitenApp->conf->force_ja_JP == TRUE) putenv("LC_ALL=ja_JP");
		if (gjitenApp->conf->force_language_c == TRUE) putenv("LANGUAGE=C");
	}

  gtk_set_locale();  

#ifdef ENABLE_NLS
  bindtextdomain(PACKAGE, GJITEN_LOCALE_DIR);
  bind_textdomain_codeset(PACKAGE, "UTF-8");
  textdomain(PACKAGE);
#endif

  gnome_program_init("gjiten", VERSION, LIBGNOMEUI_MODULE, argc, argv, 
										 GNOME_PARAM_POPT_TABLE, arg_options, 
										 GNOME_PARAM_HUMAN_READABLE_NAME, _("gjiten"), 
										 GNOME_PARAM_APP_DATADIR, GNOMEDATADIR,
										 NULL);

  if (! g_file_test (icon_path, G_FILE_TEST_EXISTS)) {
    g_warning ("Could not find %s", icon_path);
  }
  else {
    gnome_window_icon_set_default_from_file(icon_path);
  }

  /* the following is for clipboard lookup. */
  if ((gjitenApp->conf->clip_kanji_lookup == TRUE) || (gjitenApp->conf->clip_word_lookup == TRUE)) {
    if (gjitenApp->conf->clip_word_lookup) {
      gjitenApp->worddic = worddic_create();
      worddic_paste(gjitenApp->worddic);
      on_text_entered();
    }
    else {
      if (gjitenApp->conf->clip_kanji_lookup){
				clipboard_text = gtk_clipboard_wait_for_text(gtk_clipboard_get(GDK_SELECTION_PRIMARY));
				/* validate	*/
				/* FIXME: try to convert EUC-JP to UTF8 if it's non-utf8 */
				if (g_utf8_validate(clipboard_text, -1, NULL) == FALSE) {
					gjiten_print_error(_("Unable to look up kanji: NON-UTF8 string received from clipboard!\n"));
					exit(0);
				}
				else if (isKanjiChar(g_utf8_get_char(clipboard_text)) == FALSE) {
					gjiten_print_error(_("Non-kanji string received from clipboard: %s\n"), clipboard_text);
					exit(0);
				}
				else {
					if (gjitenApp->kanjidic == NULL) kanjidic_create();
					print_kanjinfo(g_utf8_get_char(clipboard_text));
					/* if (clipboard_text !	= NULL) gtk_entry_set_text(GTK_ENTRY(GTK_COMBO(combo_entry)->entry), clipboard_text); */
				}
      }
    }
  }

  if (argc > 1) {
    if (gjitenApp->conf->startkanjidic) {
      kanjidic_create();
    }
    else  
      if (gjitenApp->conf->word_to_lookup) {
				gjitenApp->worddic = worddic_create();
				gtk_entry_set_text(GTK_ENTRY(GTK_COMBO(gjitenApp->worddic->combo_entry)->entry), gjitenApp->conf->word_to_lookup);
				on_text_entered();
      }
      else if (gjitenApp->conf->kanji_to_lookup != NULL) {
				if (g_utf8_validate(gjitenApp->conf->kanji_to_lookup, -1, NULL) == FALSE) {
					gjiten_print_error(_("Unable to look up kanji: NON-UTF8 string received from clipboard!\n"));
					exit(0); /* FIXME */
				}
				else if (isKanjiChar(g_utf8_get_char(gjitenApp->conf->kanji_to_lookup)) == FALSE) {
					gjiten_print_error(_("Non-kanji string received from clipboard: %s\n"), gjitenApp->conf->kanji_to_lookup);
					exit(0); /* FIXME */
				}
				else {
					if (gjitenApp->kanjidic == NULL) kanjidic_create();
					print_kanjinfo(g_utf8_get_char(gjitenApp->conf->kanji_to_lookup));
				}
			}
			else if (!gjitenApp->conf->clip_kanji_lookup && !gjitenApp->conf->clip_word_lookup) 
				gjitenApp->worddic = worddic_create();
  }
  else {
    gjitenApp->worddic = worddic_create();
  }
	gjiten_flush_errors();
  gtk_main();
  return 0;
}
